File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        if (height.length == 0) {
            return 0;
        }

        int lo = 0, hi = height.length - 1;
        int maxL = height[lo], maxR = height[hi];
        int result  = 0;

        while (lo < hi) {
            if (maxL < maxR) {
                lo++;
                maxL = Math.max(maxL, height[lo]);
                result += Math.max(0, maxL - height[lo]);
            } else {
                hi--;
                maxR = Math.max(maxR, height[hi]);
                result += Math.max(0, maxR - height[hi]);
            }
        }

        return result;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height: return 0

        lo, hi = 0, len(height) - 1
        maxL, maxR = height[lo], height[hi]
        result = 0

        while lo < hi:
            if maxL < maxR:
                lo += 1
                maxL = max(maxL, height[lo])
                result += max(0, maxL - height[lo])
            else:
                hi -= 1
                maxR = max(maxR, height[hi])
                result += max(0, maxR - height[hi])

        return result